url29ae058134_yuriy_chumak_jatha_tgz-pJ8-org_jatha_LispJ8/src/org/jatha/compile/LispCompiler.java

```diff
@@ -0,0 +1,1678 @@
+	 * @param expr expression to compile
+	 * @param varValues  global or local variable list.
+	 * @return LispValue - generated code
+	 */
+	public LispCons compile(SECDMachine machine, LispValue expr, LispList varValues)
+			throws CompilerException
+	{
+		if (DEBUG) {
+			System.out.println("expr = " + expr);
+			System.out.println("varValues = " + varValues);
+			System.out.println("STOP = " + STOP);
+			System.out.println("NIL = " + NIL);
+			System.out.println("initial code = " + cons(STOP, NIL));
+		}
+
+		return compile(expr, varValues, cons(STOP, NIL));
+	}
+
+	// @author  Micheal S. Hewett    hewett@cs.stanford.edu
+	// @date    Sun Feb  2 19:46:35 1997
+	/**
+	 * <tt>compile</tt> takes a LISP expression, a list of
+	 * global variables, and optionally an already-generated
+	 * list of code.  It returns compiled code in a list.
+	 *
+	 * @see LispCompiler
+	 * @param expr expression to compile
+	 * @param valueList global variable list.
+	 * @param  code [optional]
+	 * @return LispValue - generated code
+	 */
+	public LispCons compile(LispValue expr, LispList valueList, LispCons code)
+			throws CompilerException
+	{
+		SECDMachine machine = f_lisp.MACHINE;
+		
+		if (DEBUG) {
+			System.out.print("\nCompile: " + expr);
+			System.out.print("\n   code: " + code);
+		}
+
+		if (expr instanceof LispAtom)
+			return compileAtom(machine, (LispAtom)expr, valueList, code);
+		else
+			return compileList(machine, (LispList)expr, valueList, code);
+	}
+
+
+	LispCons compileAtom(SECDMachine machine, LispAtom expr, LispList valueList, LispCons code)
+	{
+		if (DEBUG) {
+			System.out.print("\nCompile Atom: " + expr);
+			System.out.print(" of type " + expr.getClass().getName());
+			System.out.flush();
+		}
+
+		if (expr == T)
+			return cons(LDT, code);
+
+		// todo: remove LispConstant from code as unneeded
+		if (expr instanceof LispConstant)
+			return cons(LDC, cons(expr, code));
+
+		if (expr instanceof LispSymbol) {
+			//LispValue varIndex = index(expr, valueList);
+			//##JPG use indexAndAttributes() instead of index
+			LispValue varIdxAndAttributes = indexAndAttribute(expr, valueList);
+			LispValue paramAttribute      = car(varIdxAndAttributes);
+			LispValue varIndex            = car(cdr(varIdxAndAttributes));
+
+			if (varIndex == NIL)
+			{
+				if (!expr.specialP() && WarnAboutSpecialsP) /* Not a local variable, maybe it's global */
+					System.err.print("\n;; ** Warning - " + expr.toString() + " assumed special.\n");
+
+				return cons(LD_GLOBAL, cons(expr, code));
+			}
+			else  /* Found the symbol.  Is it bound? */
+			{
+		        //##JPG opcode LDR instead of LD for variable arguments
+		        // note : paramAttribute can only be nil or &rest
+				LispValue loadOpCode = (paramAttribute == AMP_REST) ? LDR : LD;
+				return cons(loadOpCode, cons(varIndex, code));
+			}
+		}
+		
+		return cons(LDC, cons(expr, code));
+	}
+
+	LispCons compileList(SECDMachine machine, LispList expr, LispList valueList, LispCons code)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 24. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.